### PR TITLE
fix(analyzer): False positive in rule noUndeclaredVariables #243

### PR DESCRIPTION
**Pull request by denbezrukov**
## Summary

Fix https://github.com/biomejs/biome/issues/243

We don't need to check the correct position of infer types in semantics because our parser raises an error if they are in an invalid position. #308

Thank to @strager for the explanation.

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by ematipico** [comment]
!bench_analyzer

I want to check if the workflow is fixed

**Comment by denbezrukov** [comment]
> !bench_analyzer
> 
> I want to check if the workflow is fixed 

What do you think if there is a better way to check if a type is inside a conditional branch?
We have to go to a parent and compare a node with a true type for every TsNode.
I'm wondering if we can somehow know that the next node is a true type.

**Comment by ematipico** [comment]
I think the implementation is fine as is! 

Sorry if I used your PR to trigger the benchmark. The workflows were broken for a long time, and I wanted to check if they were finally fixed.

File: crates/biome_js_semantic/src/events.rs [hunk 1]
```diff
@@ -321,8 +324,11 @@ impl SemanticEventExtractor {
                     false,
                 );
             }
-
-            _ => {}
+            _ => {
+                if let Some(node) = node.clone().cast::<AnyTsType>() {
```
Comment (Conaclos) [inline crates/biome_js_semantic/src/events.rs:328]: We could clone only when the cast is possible:
```suggestion
                if let Some(node) = AnyTsType::cast_ref(node) {
```

File: crates/biome_js_semantic/src/events.rs [hunk 2]
```diff
@@ -562,15 +587,26 @@ impl SemanticEventExtractor {
...
             | TS_INTERFACE_DECLARATION
             | TS_ENUM_DECLARATION
             | TS_TYPE_ALIAS_DECLARATION
             | TS_MODULE_DECLARATION
             | TS_EXTERNAL_MODULE_DECLARATION => {
                 self.pop_scope(node.text_range());
             }
-            _ => {}
+            _ => {
+                if let Some(node) = node.clone().cast::<AnyTsType>() {
```
Comment (Conaclos) [inline crates/biome_js_semantic/src/events.rs:598]:
```suggestion
                if let Some(node) = AnyTsType::cast_ref(node) {
```

Comment (denbezrukov) [reply crates/biome_js_semantic/src/events.rs]: Thank you, I didn't know about it =)

**Pull request merged**